src/collapse/dtrace.rs

```diff
@@ -232,38 +405,172 @@ impl Folder {
             }
         }
         // count it!
-        *self.occurrences.entry(stack_str).or_insert(0) += count;
+        self.occurrences.add(stack_str, count);
+
         // reset for the next event
         self.stack_str_size = 0;
         self.stack.clear();
     }
+}
 
-    fn finish<W: Write>(&self, mut writer: W) -> io::Result<()> {
-        let mut keys: Vec<_> = self.occurrences.iter().collect();
-        keys.sort();
-        for (key, count) in keys {
-            writeln!(writer, "{} {}", key, count)?;
+fn is_end_of_stack(line: &[u8]) -> bool {
+    enum State {
+        Start,
+        Middle,
+        End,
+    }
+    let mut state = State::Start;
+    for b in line {
+        let b = *b;
+        let c = b as char;
+        match state {
+            State::Start => {
+                if c.is_whitespace() {
+                    continue;
+                }
+                if b > 47 && b < 58 {
```

Comment:
I'd prefer to use `b > b'/' && b < b':'` here. This also deserves a comment, as it's pretty obscure why you're using that range specifically.